cmd/auth-rpc-client.go

```diff
@@ -80,13 +80,13 @@ func newAuthRPCClient(config authConfig) *AuthRPCClient {
 
 // Login - a jwt based authentication is performed with rpc server.
 func (authClient *AuthRPCClient) Login() (err error) {
-	authClient.Lock()
-	defer authClient.Unlock()
-
+	authClient.RLock()
 	// Return if already logged in.
 	if authClient.authToken != "" {
+		authClient.RUnlock()
 		return nil
 	}
+	authClient.RUnlock()
```

Comment:
Yes you are right there is no way to prevent this in the current code style, Login() is sort of mixed together with Login() as a separate function and also to be part of the call() it calls for racy situations. I have kept the code as is from previous work and just added RLock() for reading auth token.

Another idea that i explored which works is using a factory pattern for Login. By creating a loginFactory 

```
// loginFactory a JWT based authentication is performed with rpc server.
type loginFactory func() (token string, err error)

func newLoginFactory(rpcClient *RPCClient, config authConfig) loginFactory {
        var authToken string
        var mutex = &sync.Mutex{}
        return func() (string, error) {
                // Login should be attempted one at a time.
                //
                // The reason for large region lock here is
                // to avoid two simultaneous login attempts
                // racing over each other.
                //
                // #1 Login() gets the lock proceeds to login.
                // #2 Login() waits for the unlock to happen
                //    after login in #1.
                // #1 Successfully completes login saves the
                //    newly acquired token.
                // #2 Successfully gets the lock and proceeds,
                //    but since we have acquired the token
                //    already the call quickly returns.
                mutex.Lock()
                defer mutex.Unlock()

                fmt.Println("AuthToken Set", authToken != "")
                if authToken != "" {
                        return authToken, nil
                }

                // Login to authenticate and acquire a new auth token.
                var (
                        loginMethod = config.serviceName + loginMethodName
                        loginArgs   = LoginRPCArgs{
                                Username:    config.accessKey,
                                Password:    config.secretKey,
                                Version:     Version,
                                RequestTime: UTCNow(),
                        }
                        loginReply = LoginRPCReply{}
                )

                if err := rpcClient.Call(loginMethod, &loginArgs, &loginReply); err != nil {
                        return "", err
                }

                authToken = loginReply.AuthToken

                // Success.
                return authToken, nil

        }
}
```

But won't make this change for now. We will just use RLock() for authToken reading instead.